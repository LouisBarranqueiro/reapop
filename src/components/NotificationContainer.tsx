import React, {ReactNode, useEffect, useState} from 'react'
import {Timer} from '../utils'
import {Notification as NotificationType} from '../reducers/notifications/types'
import {DismissNotification} from './NotificationsSystem'

type Props = {
    notification: NotificationType
    dismissNotification: DismissNotification
    children?: ReactNode
}

const NotificationContainer = (props: Props) => {
    const {notification, dismissNotification} = props
    const {dismissAfter, onAdd, onDismiss} = notification
    const [timer, setTimer] = useState<Timer | null>(null)

    useEffect(() => {
        if (onAdd) {
            onAdd()
        }
        return () => {
            if (onDismiss) {
                onDismiss()
            }
        }
    }, [])

    useEffect(() => {
        if (!timer && dismissAfter && dismissAfter > 0) {
            const timer = new Timer(dismissAfter, () => dismissNotification(notification.id))
            timer.resume()
            setTimer(timer)
        } else if (timer && !dismissAfter) {
            timer.pause()
            setTimer(null)
        }
    }, [dismissAfter])

    return (
        <div
            data-testid="timed-notification"
            onMouseEnter={timer ? () => timer.pause() : undefined}
            onMouseLeave={timer ? () => timer.resume() : undefined}
        >
            {props.children}
        </div>
    )
}

export default NotificationContainer
