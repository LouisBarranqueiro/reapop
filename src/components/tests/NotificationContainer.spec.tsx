import React from 'react'
import pretty from 'pretty'
import {render, fireEvent, act} from '@testing-library/react'

import {Notification, POSITIONS, STATUSES} from '../..'
import NotificationContainer from '../NotificationContainer'

describe('<NotificationContainer/>', () => {
    const baseNotification: Notification = {
        id: '1',
        message: 'hello world!',
        position: POSITIONS.topLeft,
        status: STATUSES.none,
        buttons: [],
    }

    beforeEach(() => {
        jest.clearAllMocks()
    })

    afterAll(() => {
        jest.resetAllMocks()
    })

    it('should call `onAdd` function when notification is displayed', () => {
        const notification = {
            ...baseNotification,
            onAdd: jest.fn(),
        }
        render(<NotificationContainer notification={notification} dismissNotification={jest.fn()} />)
        expect(notification.onAdd).toHaveBeenNthCalledWith(1)
    })

    it('should call `onDismiss` function when notification is dimissed', () => {
        const notification = {
            ...baseNotification,
            onDismiss: jest.fn(),
        }
        const {unmount} = render(<NotificationContainer notification={notification} dismissNotification={jest.fn()} />)
        unmount()
        expect(notification.onDismiss).toHaveBeenNthCalledWith(1)
    })

    it('should pause/resume timer when mouse is entering/leaving the notification', (done) => {
        const notification = {
            ...baseNotification,
            dismissAfter: 4,
        }
        const dismissNotificationSpy = jest.fn()
        const {getByTestId} = render(
            <NotificationContainer notification={notification} dismissNotification={dismissNotificationSpy} />
        )
        const element = getByTestId('timed-notification')

        act(() => {
            fireEvent.mouseEnter(element)
        })

        setTimeout(() => {
            // 1. Notification was still not removed because mouse is hovering it.
            expect(dismissNotificationSpy).not.toHaveBeenCalled()

            // 2. Mouse leaves the notification so the function used to dismiss it should be called soon.
            fireEvent.mouseLeave(element)

            setTimeout(() => {
                // 3. Verify the function used to dismiss the notification has been called.
                expect(dismissNotificationSpy).toHaveBeenNthCalledWith(1, notification.id)
                done()
            }, notification.dismissAfter)
        }, notification.dismissAfter)
    })

    it('should pause timer when notification `dismissAfter` timeout is removed', (done) => {
        const dismissAfter = 4
        const notification = {
            ...baseNotification,
            dismissAfter,
        }
        const dismissNotificationSpy = jest.fn()
        const {rerender} = render(
            <NotificationContainer notification={notification} dismissNotification={dismissNotificationSpy} />
        )

        // remove the `dismissAfter` timeout
        rerender(
            <NotificationContainer
                notification={{...notification, dismissAfter: undefined}}
                dismissNotification={dismissNotificationSpy}
            />
        )

        setTimeout(() => {
            expect(dismissNotificationSpy).not.toHaveBeenCalled()
            done()
        }, dismissAfter)
    })

    it('should resume timer when notification `dismissAfter` timeout is added', (done) => {
        const dismissAfter = 4
        const dismissNotificationSpy = jest.fn()
        const {rerender} = render(
            <NotificationContainer notification={baseNotification} dismissNotification={dismissNotificationSpy} />
        )

        setTimeout(() => {
            expect(dismissNotificationSpy).not.toHaveBeenCalled()
        }, dismissAfter)

        // add the `dismissAfter` timeout
        rerender(
            <NotificationContainer
                notification={{...baseNotification, dismissAfter}}
                dismissNotification={dismissNotificationSpy}
            />
        )

        setTimeout(() => {
            expect(dismissNotificationSpy).toHaveBeenNthCalledWith(1, baseNotification.id)
            done()
        }, dismissAfter + 1)
    })

    it('should render a container', () => {
        const {container} = render(
            <NotificationContainer notification={baseNotification} dismissNotification={jest.fn()}>
                <div>hello world!</div>
            </NotificationContainer>
        )
        expect(pretty(container.innerHTML)).toMatchSnapshot()
    })
})
