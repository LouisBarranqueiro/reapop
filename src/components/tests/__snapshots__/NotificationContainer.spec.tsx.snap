// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationContainer/> should render a container 1`] = `
"<div data-testid=\\"timed-notification\\">
  <div>hello world!</div>
</div>"
`;
